ecosystem/ffmpeg_plugin/kahawai_enc.c

```diff
@@ -0,0 +1,366 @@
+      -1 here because some times we cannot get all notifications from MTL
+    */
+    if (s->frame_tx_completed == (s->frame_counter - 1)) {
+      s->frame_tx_completed = 0;
+      s->frame_counter = 0;
+      s->tx_completed = true;
+    }
+  }
+
+  return 0;
+}
+
+static int kahawai_write_header(AVFormatContext *ctx) {
+  KahawaiMuxerContext *s = ctx->priv_data;
+  AVStream *st = NULL;
+
+  int packet_size = 0;
+
+  int ret = 0;
+
+  // struct mtl_init_params param;
+  struct st20p_tx_ops ops_tx;
+
+  av_log(ctx, AV_LOG_VERBOSE, "kahawai_write_header triggered\n");
+
+  memset(&ops_tx, 0, sizeof(ops_tx));
+
+  if ((NULL == s->port) || (strlen(s->port) > MTL_PORT_MAX_LEN)) {
+    av_log(ctx, AV_LOG_ERROR, "Invalid port info\n");
+    return AVERROR(EINVAL);
+  }
+  ops_tx.port.num_port = 1;
+  strncpy(ops_tx.port.port[MTL_PORT_P], s->port, MTL_PORT_MAX_LEN);
+
+  if (NULL == s->dst_addr) {
+    av_log(ctx, AV_LOG_ERROR, "Invalid destination IP address\n");
+    return AVERROR(EINVAL);
+  } else if (sscanf(s->dst_addr, "%hhu.%hhu.%hhu.%hhu",
+                    &ops_tx.port.dip_addr[MTL_PORT_P][0],
+                    &ops_tx.port.dip_addr[MTL_PORT_P][1],
+                    &ops_tx.port.dip_addr[MTL_PORT_P][2],
+                    &ops_tx.port.dip_addr[MTL_PORT_P][3]) != MTL_IP_ADDR_LEN) {
+    av_log(ctx, AV_LOG_ERROR, "Failed to parse destination IP address: %s\n",
+           s->dst_addr);
+    return AVERROR(EINVAL);
+  }
+
+  if ((s->udp_port < 0) || (s->udp_port > 0xFFFF)) {
+    av_log(ctx, AV_LOG_ERROR, "Invalid UDP port: %d\n", s->udp_port);
+    return AVERROR(EINVAL);
+  }
+  ops_tx.port.udp_port[MTL_PORT_P] = s->udp_port;
+
+  ops_tx.width = s->width = ctx->streams[0]->codecpar->width;
+  ops_tx.height = s->height = ctx->streams[0]->codecpar->height;
+
+  s->pixel_format = ctx->streams[0]->codecpar->format;
+  if (s->pixel_format == AV_PIX_FMT_NONE) {
+    av_log(ctx, AV_LOG_ERROR, "No such pixel format: %s.\n",
+           av_pix_fmt_desc_get(s->pixel_format));
+    return AVERROR(EINVAL);
+  } else if (s->pixel_format != AV_PIX_FMT_RGB24) {
+    av_log(ctx, AV_LOG_ERROR, "Only rgb24 are supported\n");
+    return AVERROR(EINVAL);
+  }
+
+  ops_tx.transport_fmt = ST20_FMT_RGB_8BIT;
+  ops_tx.input_fmt = ST_FRAME_FMT_RGB8;
+
+  s->framerate = ctx->streams[0]->avg_frame_rate;
+  if (ops_tx.fps = get_fps_table(s->framerate) == ST_FPS_MAX) {
+    av_log(ctx, AV_LOG_ERROR, "Frame rate %f is not supported\n",
+           av_q2d(s->framerate));
+    return AVERROR(EINVAL);
+  }
+
+  // Create device
+  if (!kahawai_get_handle()) {
+    s->dev_handle = kahawai_init(s->port, s->local_addr, s->udp_port,
+                                 s->session_cnt, 0, NULL);
+    if (!s->dev_handle) {
+      av_log(ctx, AV_LOG_ERROR, "mtl_init failed\n");
+      return AVERROR(EIO);
+    }
+    kahawai_set_handle(s->dev_handle);
+    av_log(ctx, AV_LOG_VERBOSE, "mtl_init finished: st_handle 0x%" PRIx64 "\n",
```

Comment:
nit: why not use %p to print the handle?

Reply:
Changed to %p format specifier to print the pointer of handle